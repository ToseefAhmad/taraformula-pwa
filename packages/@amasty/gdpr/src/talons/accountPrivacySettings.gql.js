import { gql } from '@apollo/client';

export const GET_ACCOUNT_SETTINGS = gql`
  query amGetGdprAccountAction {
    amGetGdprAccountAction {
      title
      content
      hasCheckbox
      checkboxText
      hidePassword
      needPassword
      submitText
      action
      actionCode
    }
  }
`;

export const ACCEPT_GDPR_POLICY = gql`
  mutation acceptAmGdprPolicy($policy_version: String!) {
    acceptAmGdprPolicy(input: { policy_version: $policy_version }) {
      error
      message
    }
  }
`;

export const ANONIMISE_CUSTOMER_DATA = gql`
  mutation anonymiseAmGdprCustomer($password: String!) {
    anonymiseAmGdprCustomer(input: { password: $password }) {
      error
      message
    }
  }
`;

export const DOWNLOAD_CUSTOMER_DATA = gql`
  mutation amGetGdprCustomerPersonalData($password: String!) {
    amGetGdprCustomerPersonalData(input: { password: $password }) {
      label
      value
    }
  }
`;

export const DELETE_CUSTOMER_ACCOUNT = gql`
  mutation requestAmGdprAccountDeletion($password: String!) {
    requestAmGdprAccountDeletion(input: { password: $password }) {
      error
      message
    }
  }
`;

export default {
  getAccountSettings: GET_ACCOUNT_SETTINGS,
  acceptGdprPolicyMutation: ACCEPT_GDPR_POLICY,
  anonimiseCustomerDataMutation: ANONIMISE_CUSTOMER_DATA,
  downloadCustomerDataMutation: DOWNLOAD_CUSTOMER_DATA,
  deleteCustomerAccountMutation: DELETE_CUSTOMER_ACCOUNT
};
