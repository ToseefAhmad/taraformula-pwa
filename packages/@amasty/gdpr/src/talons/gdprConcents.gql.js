import { gql } from '@apollo/client';

export const GET_GDPR_CONSENTS = gql`
  query amGetGdprConsents {
    amGetGdprConsents {
      consent_id
      name
      consent_code
      is_required
      hide_the_consent_after_user_left_the_consent
      consent_location
      link_type
      cms_page_id
      consent_text
      visibility
      has_agreement
      sort_order
    }
  }
`;

export const GET_GDPR_POLICY_TEXT = gql`
  query amGetGdprPolicyText($id: Int!) {
    amGetGdprPolicyText(id: $id) {
      title
      content
      need_show
      version
      version_changed
    }
  }
`;

export const GIVE_GDPR_CONSENT = gql`
  mutation giveAmGdprConsent($consents: [AmGdprConsentInput!]!) {
    giveAmGdprConsent(input: { consents: $consents }) {
      error
      message
    }
  }
`;

export const ACCEPT_GDPR_POLICY = gql`
  mutation acceptAmGdprPolicy($policy_version: String!) {
    acceptAmGdprPolicy(input: { policy_version: $policy_version }) {
      error
      message
    }
  }
`;

export default {
  getGdprConsents: GET_GDPR_CONSENTS,
  getGdprPolicyText: GET_GDPR_POLICY_TEXT,
  giveGdprConsentMutation: GIVE_GDPR_CONSENT,
  acceptGdprPolicyMutation: ACCEPT_GDPR_POLICY
};
