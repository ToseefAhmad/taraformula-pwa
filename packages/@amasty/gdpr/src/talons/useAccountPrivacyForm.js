import { useCallback, useState } from 'react';
import { useMutation } from '@apollo/client';
import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './accountPrivacySettings.gql.js';

export const useAccountPrivacyForm = (props = {}) => {
  const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
  const {
    deleteCustomerAccountMutation,
    downloadCustomerDataMutation,
    anonimiseCustomerDataMutation
  } = operations;

  const [isShowPassword, setIsShowPassword] = useState(!props.hidePassword);
  const [displayError, setDisplayError] = useState(false);
  const [formSubmitMessage, setFormSubmitMessage] = useState('');
  const [customerData, setCustomerData] = useState([]);
  const [deleteAccount, { error: deleteActionError }] = useMutation(
    deleteCustomerAccountMutation,
    {
      fetchPolicy: 'no-cache'
    }
  );
  const [anonymiseData, { error: anonymiseDataError }] = useMutation(
    anonimiseCustomerDataMutation,
    {
      fetchPolicy: 'no-cache'
    }
  );
  const [loadCustomerData, { error: customerDataError }] = useMutation(
    downloadCustomerDataMutation,
    {
      fetchPolicy: 'no-cache'
    }
  );

  const togglePassword = useCallback(() => {
    setIsShowPassword(!isShowPassword);
  }, [setIsShowPassword, isShowPassword]);

  const formSubmit = useCallback(
    async form => {
      try {
        if (props.actionCode === 'delete') {
          const { data } = await deleteAccount({
            variables: {
              password: form.password
            }
          });

          const { requestAmGdprAccountDeletion: deletionData } = data || {};
          setFormSubmitMessage(deletionData.message);
        }

        if (props.actionCode === 'anonymize') {
          const { data } = await anonymiseData({
            variables: {
              password: form.password
            }
          });

          const { anonymiseAmGdprCustomer: anonymiseGdprData } = data || {};
          setFormSubmitMessage(anonymiseGdprData.message);
        }

        if (props.actionCode === 'download') {
          const { data } = await loadCustomerData({
            variables: {
              password: form.password
            }
          });

          const { amGetGdprCustomerPersonalData: customerPersonalData } =
            data || {};
          setCustomerData(customerPersonalData);
        }
      } catch (error) {
        setDisplayError(true);
      }
    },
    [
      props,
      deleteAccount,
      setFormSubmitMessage,
      anonymiseData,
      loadCustomerData,
      setCustomerData
    ]
  );

  const errors = displayError
    ? [deleteActionError, anonymiseDataError, customerDataError]
    : [];

  return {
    customerData,
    formSubmitMessage,
    errors,
    togglePassword,
    isShowPassword,
    handleSubmit: formSubmit
  };
};
