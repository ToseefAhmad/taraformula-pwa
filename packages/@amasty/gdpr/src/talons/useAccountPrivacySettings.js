import { useEffect } from 'react';
import { useQuery } from '@apollo/client';
import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useHistory } from 'react-router-dom';
import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './accountPrivacySettings.gql.js';

export const useAccountPrivacySettings = (props = {}) => {
  const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
  const { getAccountSettings } = operations;
  const [{ isSignedIn }] = useUserContext();
  const history = useHistory();

  const { loading, error, data } = useQuery(getAccountSettings, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    skip: !isSignedIn
  });

  // If the user is no longer signed in, redirect to the home page.
  useEffect(() => {
    if (!isSignedIn) {
      history.push('/');
    }
  }, [history, isSignedIn]);

  const { amGetGdprAccountAction: settings } = data || {};

  return {
    settingsLoading: loading,
    error,
    settings
  };
};
