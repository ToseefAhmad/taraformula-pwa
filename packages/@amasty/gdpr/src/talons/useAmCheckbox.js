import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './gdprConcents.gql.js';
import { useMutation, useQuery } from '@apollo/client';
import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useCallback, useEffect, useRef, useState } from 'react';

export const useAmCheckbox = (props = {}) => {
  const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
  const checkedConsents = useRef([]);
  const { getGdprConsents, giveGdprConsentMutation } = operations;
  const defaultFetchPolicy = 'cache-first';
  const [fetchPolicy, setFetchPolicy] = useState(defaultFetchPolicy);
  const [{ isSignedIn }] = useUserContext();

  const [applyCheckbox] = useMutation(giveGdprConsentMutation, {
    fetchPolicy: 'no-cache'
  });

  const acceptRegistrationConsents = useCallback(async () => {
    try {
      await applyCheckbox({
        variables: {
          consents: checkedConsents.current
        },
        refetchQueries: [{ query: getGdprConsents }],
        awaitRefetchQueries: true
      });
    } catch (error) {
      if (process.env.NODE_ENV === 'development') {
        console.error(error);
      }
    }
  }, [checkedConsents, applyCheckbox, getGdprConsents]);

  useEffect(() => {
    setFetchPolicy('network-only');

    if (isSignedIn && checkedConsents.current.length) {
      acceptRegistrationConsents();
    }
  }, [isSignedIn, acceptRegistrationConsents]);

  const { loading, error, data } = useQuery(getGdprConsents, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: fetchPolicy
  });
  const { amGetGdprConsents: settings } = data || {};

  useEffect(() => {
    if (fetchPolicy !== defaultFetchPolicy) {
      setFetchPolicy(defaultFetchPolicy);
    }
  }, [fetchPolicy, setFetchPolicy, defaultFetchPolicy]);

  return {
    settingsLoading: loading,
    error,
    settings,
    getGdprConsents,
    checkedConsents
  };
};
