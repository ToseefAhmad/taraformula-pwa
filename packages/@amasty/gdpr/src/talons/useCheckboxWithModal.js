import { useLazyQuery, useMutation } from '@apollo/client';
import { useCallback, useEffect, useState } from 'react';
import { useFieldApi } from 'informed';
import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './gdprConcents.gql.js';
import { useAmGdprContext } from '../context';
import { getCheckedConsentArray, parseLink } from '../../utils/index';

export const useCheckboxWithModal = (props = {}) => {
  const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
  const { giveGdprConsentMutation, getGdprPolicyText } = operations;
  const {
    checkboxName,
    consentCode,
    consentFrom,
    hasAgreement,
    checkboxLabel
  } = props;
  const { checkedConsents, getGdprConsents } = useAmGdprContext();
  const [formSubmitMessage, setFormSubmitMessage] = useState('');
  const [displayError, setDisplayError] = useState(false);
  const [isChecked, setIsChecked] = useState(false);
  const [isShowModal, setIsShowModal] = useState(false);
  const checkboxApi = useFieldApi(checkboxName);
  const [applyCheckbox, { error: applyCheckboxError }] = useMutation(
    giveGdprConsentMutation,
    {
      fetchPolicy: 'no-cache'
    }
  );
  const [loadQuery, { data: modalData }] = useLazyQuery(getGdprPolicyText, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first'
  });
  const label = parseLink(checkboxLabel);

  useEffect(() => {
    if (hasAgreement) {
      checkboxApi.setValue(true);
      setIsChecked(true);
    }
  }, [hasAgreement, checkboxApi]);

  const handleCancel = useCallback(() => {
    setIsShowModal(false);
  }, [setIsShowModal]);

  const handleClick = useCallback(async () => {
    if (consentFrom === 'registration') {
      getCheckedConsentArray(
        checkedConsents,
        consentFrom,
        consentCode,
        !isChecked
      );
    } else {
      try {
        const { data } = await applyCheckbox({
          variables: {
            consents: [
              {
                consent_from: consentFrom,
                consent_code: consentCode,
                has_agreement: !isChecked
              }
            ]
          },
          refetchQueries: [{ query: getGdprConsents }],
          awaitRefetchQueries: true
        });

        const { giveAmGdprConsent: applyCheckboxData } = data || {};
        setFormSubmitMessage(applyCheckboxData.message);
      } catch (error) {
        setDisplayError(true);
      }
    }
    setIsChecked(prev => !prev);
  }, [
    applyCheckbox,
    consentFrom,
    consentCode,
    setFormSubmitMessage,
    isChecked,
    setIsChecked,
    getGdprConsents,
    checkedConsents
  ]);

  const handleSubmit = useCallback(async () => {
    setIsShowModal(false);
    if (!isChecked) {
      checkboxApi.setValue(true);
      setIsChecked(true);
      await handleClick();
    }
  }, [setIsShowModal, checkboxApi, setIsChecked, handleClick, isChecked]);

  const getModalData = useCallback(
    id => {
      loadQuery({
        variables: { id }
      });

      setIsShowModal(true);
    },
    [loadQuery]
  );

  const { amGetGdprPolicyText: modalContent } = modalData || {};
  const errors = displayError ? [applyCheckboxError] : [];

  return {
    isShowModal,
    handleSubmit,
    handleCancel,
    modalContent,
    getModalData,
    handleClick,
    errors,
    formSubmitMessage,
    label
  };
};
