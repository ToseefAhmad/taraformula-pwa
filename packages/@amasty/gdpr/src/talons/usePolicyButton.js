import { useCallback, useState } from 'react';
import { useLazyQuery, useMutation } from '@apollo/client';
import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './gdprConcents.gql.js';

export const usePolicyButton = (props = {}) => {
  const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
  const { acceptGdprPolicyMutation, getGdprPolicyText } = operations;
  const [isShowMode, setIsShowMode] = useState(false);

  const [loadQuery, { data: modalData }] = useLazyQuery(getGdprPolicyText, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first'
  });
  const { amGetGdprPolicyText: modalContent } = modalData || {};
  const [acceptPolicy] = useMutation(acceptGdprPolicyMutation, {
    fetchPolicy: 'no-cache'
  });

  const handleCancel = useCallback(() => {
    setIsShowMode(false);
  }, [setIsShowMode]);

  const showUpdateMode = useCallback(() => {
    loadQuery({
      variables: { id: 0 }
    });
    setIsShowMode(true);
  }, [setIsShowMode, loadQuery]);

  const handleSubmit = useCallback(async () => {
    try {
      await acceptPolicy({
        variables: {
          policy_version: modalContent.version
        }
      });
    } catch (error) {
      if (process.env.NODE_ENV === 'development') {
        console.error(error);
      }
    }
    setIsShowMode(false);
  }, [setIsShowMode, acceptPolicy, modalContent]);

  return {
    modalContent,
    isShowMode,
    showUpdateMode,
    handleCancel,
    handleSubmit
  };
};
