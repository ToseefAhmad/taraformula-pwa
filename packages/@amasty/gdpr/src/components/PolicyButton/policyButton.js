import React, { Fragment, Suspense } from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { FormattedMessage } from 'react-intl';
import defaultClasses from '../AccountPrivacySettings/accountPrivacySettings.css';
import PrivacyModal from '../PrivacyModal';
import LinkButton from '@magento/venia-ui/lib/components/LinkButton';
import { usePolicyButton } from '../../talons/usePolicyButton';

const PolicyButton = props => {
  const classes = mergeClasses(defaultClasses, props.classes);
  const {
    isShowMode,
    showUpdateMode,
    handleCancel,
    handleSubmit,
    modalContent
  } = usePolicyButton();

  return (
    <Fragment>
      <div className={classes.item}>
        <h3 className={classes.title}>{props.item.title}</h3>
        <LinkButton type="button" onClick={showUpdateMode}>
          <FormattedMessage
            id={`privacy.submitButton${props.index}`}
            defaultMessage={'Read Privacy Policy'}
          />
        </LinkButton>
      </div>
      <Suspense fallback={null}>
        <PrivacyModal
          isOpen={isShowMode}
          onCancel={handleCancel}
          onSubmit={handleSubmit}
          content={modalContent}
        />
      </Suspense>
    </Fragment>
  );
};

export default PolicyButton;
