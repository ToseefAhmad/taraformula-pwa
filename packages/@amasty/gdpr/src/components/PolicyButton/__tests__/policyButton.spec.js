import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import PolicyButton from '../policyButton';
import { usePolicyButton } from '../../../talons/usePolicyButton';

jest.mock('../../../talons/usePolicyButton', () => {
  const usePolicyButton = jest.fn(() => {});

  return { usePolicyButton };
});
jest.mock('react-intl', () => ({
  FormattedMessage: props => (
    <div componentName="Formatted Message Component" {...props} />
  )
}));
jest.mock('../../PrivacyModal/privacyModal', () => 'PrivacyModal');
jest.mock('@magento/venia-ui/lib/components/LinkButton', () => 'LinkButton');

test('renders default policyButton', () => {
  usePolicyButton.mockReturnValue({
    isShowMode: true
  });

  const tree = createTestInstance(
    <PolicyButton item={{ title: 'title' }} index={0} />
  );
  expect(tree.toJSON()).toMatchSnapshot();
});
