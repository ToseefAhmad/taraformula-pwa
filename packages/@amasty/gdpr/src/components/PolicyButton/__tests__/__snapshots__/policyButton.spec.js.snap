// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default policyButton 1`] = `
Array [
  <div>
    <h3>
      title
    </h3>
    <LinkButton
      type="button"
    >
      <div
        componentName="Formatted Message Component"
        defaultMessage="Read Privacy Policy"
        id="privacy.submitButton0"
      />
    </LinkButton>
  </div>,
  <PrivacyModal
    isOpen={true}
  />,
]
`;
